saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/Saml2AuthenticationRequestContext.java

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.saml2.provider.service.authentication;
+
+import org.springframework.security.saml2.credentials.Saml2X509Credential;
+import org.springframework.util.Assert;
+
+import java.util.Collection;
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+
+import static java.util.Collections.emptyMap;
+import static org.springframework.util.StringUtils.hasText;
+
+/**
+ * Data holder for information required to create an {@code AuthNRequest}
+ * to be sent from the service provider to the identity provider
+ * https://www.oasis-open.org/committees/download.php/35711/sstc-saml-core-errata-2.0-wd-06-diff.pdf (line 2031)
+ *
+ * @see {@link Saml2AuthenticationRequestFactory}
+ * @since 5.3
+ */
+public final class Saml2AuthenticationRequestContext {
+	private final String issuer;
+	private final List<Saml2X509Credential> credentials;
+	private final String destination;
+	private final String assertionConsumerServiceUrl;
+	private final String relayState;
+	private final Map<String, String> additionalParameters;
+
+	private Saml2AuthenticationRequestContext(
+			String issuer,
+			String destination,
+			String assertionConsumerServiceUrl,
+			List<Saml2X509Credential> credentials,
+			String relayState,
+			Map<String, String> additionalParameters) {
```

Comment:
We can actually remove `additionalParameters` all together, and add it upon request, as we have it as a _nice to have_ rather than driven by an actual requirement.